src/goto-symex/symex_main.cpp

```diff
@@ -541,3 +562,145 @@ void goto_symext::symex_step(
+
+  value_setst::valuest value_set_elements;
+  original_value_set.get_value_set(*symbol_expr, value_set_elements, ns);
+
+  std::vector<exprt> delete_from_jump_taken_value_set;
+  std::vector<exprt> delete_from_jump_not_taken_value_set;
+  delete_from_jump_taken_value_set.reserve(value_set_elements.size());
+  delete_from_jump_not_taken_value_set.reserve(value_set_elements.size());
+
+  // Try evaluating the condition with the symbol replaced by a pointer to each
+  // one of its possible values in turn. If that leads to a true for some
+  // value_set_element then we can delete it from the value set that will be
+  // used if the condition is false, and vice versa.
+  for(const auto &value_set_element : value_set_elements)
+  {
+    if(value_set_element.id() == ID_unknown)
+    {
+      continue;
+    }
+
+    const bool exclude_null_derefs = false;
+    value_set_dereferencet::valuet possible_value =
+      value_set_dereferencet::build_reference_to(
+        value_set_element,
+        *symbol_expr,
+        exclude_null_derefs,
+        language_mode,
+        ns);
+
+    if(possible_value.ignore)
+    {
+      continue;
+    }
+
+    exprt replacement_expr =
+      possible_value.value.is_nil()
+        ? static_cast<exprt>(null_pointer_exprt{symbol_type})
+        : static_cast<exprt>(address_of_exprt{possible_value.value});
+
+    exprt modified_condition(condition);
+
+    address_of_aware_replace_symbolt replace_symbol{};
+    replace_symbol.insert(*symbol_expr, replacement_expr);
+    replace_symbol(modified_condition);
+
+    // This do_simplify() is needed for the following reason: if condition is
+    // `*p == a` and we replace `p` with `&a` then we get `*&a == a`. Suppose
+    // our constant propagation knows that `a` is `. Without this call to
+    // do_simplify(), state.rename() turns this into `*&a == 1` (because
+    // rename() doesn't do constant propagation inside addresses), which
+    // do_simplify() turns into `a == 1`, which cannot be evaluated as true
+    // without another round of constant propagation.
+    // It would be sufficient to replace this call to do_simplify() with
+    // something that just replaces `*&x` with `x` whenever it finds it.
+    do_simplify(modified_condition);
+
+    const bool record_events = state.record_events;
+    state.record_events = false;
+    modified_condition = state.rename(std::move(modified_condition), ns).get();
+    state.record_events = record_events;
+
+    do_simplify(modified_condition);
+
+    if(jump_taken_value_set && modified_condition.is_false())
+    {
+      delete_from_jump_taken_value_set.emplace_back(value_set_element);
+    }
+    else if(jump_not_taken_value_set && modified_condition.is_true())
+    {
+      delete_from_jump_not_taken_value_set.emplace_back(value_set_element);
+    }
+  }
+  if(jump_taken_value_set && !delete_from_jump_taken_value_set.empty())
+  {
+    value_sett::entryt *entry = const_cast<value_sett::entryt *>(
```

Comment:
`const_cast`s are concerning; add a method to `value_sett`?

Reply:
I could add a method to `value_sett` that gets `get_entry_for_symbol`, `const_cast`s the result and then erases the whole vector from it. That's just moving the `const_cast` around - is that what you meant? If you want me to get rid of the `const_cast` altogether then that will be quite a bit more work.

Reply:
What is the reason that we should tolerate a `const_cast` here?

Reply:
I've now made a non-const version of `get_entry_for_symbol` to go with the const version. The const version uses a `const_cast` to call the non-const version. I hope this is okay to you, @tautschnig . This was the approach that @smowton suggested.

Reply:
In brief, the problem is that `get_entry_for_symbol` needs to be called in both const and non-const contexts, and we want to avoid duplicating its logic.

Reply:
To be honest, I'd go to pretty great length to avoid a `const_cast`, including literal copies of code.